Spec additions for encryption (#12162)
apache/iceberg: format/spec.md

| _optional_ | _optional_ | _optional_ | **`schema-id`**              | ID of the table's current schema when the snapshot was created                                                                     |
|            |            | _optional_ | **`first-row-id`**           | The first `_row_id` assigned to the first row in the first data file in the first manifest, see [Row Lineage](#row-lineage)        |
|            |            | _optional_ | **`added-rows`**             | Sum of the [`added_rows_count`](#manifest-lists) from all manifests added in this snapshot. Required if [Row Lineage](#row-lineage) is enabled | 

> |            |            | _optional_ | **`key-metadata-id`**        | ID of the manifest list key metadata in the encryption-keys table ([Encryption Keys](#encryption-keys))                            |

Author: maybe we don't need this addition.. We can use the snapshot-id to reference the snapshot's key-metadata in the encryption-keys table.

| _optional_ | _optional_ | _optional_ | **`schema-id`**              | ID of the table's current schema when the snapshot was created                                                                     |
|            |            | _optional_ | **`first-row-id`**           | The first `_row_id` assigned to the first row in the first data file in the first manifest, see [Row Lineage](#row-lineage)        |
|            |            | _optional_ | **`added-rows`**             | Sum of the [`added_rows_count`](#manifest-lists) from all manifests added in this snapshot. Required if [Row Lineage](#row-lineage) is enabled | 
|            |            | _optional_ | **`encrypted-key-metadata`** | Base64-encoded key metadata of the manifest list file in an encrypted table. The key metadata is encrypted by a metadata encryption key before encoding |

> |            |            | _optional_ | **`key-metadata-key-id`**    | The ID of the encryption key that encrypts the manifest list key metadata                                                      |

Reviewer: After my chat with Russell, I think I would prefer to have all keys in table metadata separate from snapshot. That leaves a lot of flexibility and only requires adding a `key-id` here.

| _optional_ | _required_ | _required_ | **`summary`**                | A string map that summarizes the snapshot changes, including `operation` as a _required_ field (see below)                         |
| _optional_ | _optional_ | _optional_ | **`schema-id`**              | ID of the table's current schema when the snapshot was created                                                                     |
|            |            | _required_ | **`first-row-id`**           | The first `_row_id` assigned to the first row in the first data file in the first manifest, see [Row Lineage](#row-lineage)        |

> |            |            | _optional_ | **`key-id`**                 | ID of the encryption key that encrypts the manifest list key metadata |

Reviewer 2: would it make sense to name this `encryption-key-id` to make it more obvious that this is about encryption

Reviewer: I'm not sure what other key it would be. I think it's pretty clear.

|            |            | _optional_ | **`row-lineage`**           | A boolean, defaulting to false, setting whether or not to track the creation and updates to rows in the table. See [Row Lineage](#row-lineage).                                                                                                                                                                                                                                                  |
|            |            | _optional_ | **`next-row-id`**           | A `long` higher than all assigned row IDs; the next snapshot's `first-row-id`. See [Row Lineage](#row-lineage).                                                                                                                                                                                                                                                                                  |

> |            |            | _optional_ | **`current-key-id`**        | ID of the encryption key that encrypts the manifest list key metadata.                                                                                                                                                                                                                                                                                  |

Reviewer: I'm not sure I agree with using "current" here. I don't see how this key ID would be "current" -- it is either the key for the manifest list or it is not. How about using just `key-id`?

Author: sounds good, changing to `key-id`

Reviewer: Sorry, my comment was based on thinking this was a snapshot's key ID because of the description. This was actually in table metadata.

Can we trust table metadata to provide a key ID to use? I would probably have the encryption manager select the key automatically -- the one that is available from KMS.

Author: Agreed, we can have this handled by the encryption manager.

|            |            | _optional_ | **`row-lineage`**           | A boolean, defaulting to false, setting whether or not to track the creation and updates to rows in the table. See [Row Lineage](#row-lineage).                                                                                                                                                                                                                                                  |
|            |            | _optional_ | **`next-row-id`**           | A `long` higher than all assigned row IDs; the next snapshot's `first-row-id`. See [Row Lineage](#row-lineage).                                                                                                                                                                                                                                                                                  |

> |            |            | _optional_ | **`key-cache`**             | A list of encryption keys (key-id/key-wrap pairs), used to encrypt the manifest list file key metadata. See [Snapshot](#key-metadata-key-id).                                                                                                                                                                                                                                                    |

Reviewer 3: I was wondering, if we really need multiple keys here. This would allow folks to use old keys after a key rotation but is that something we really want to enable? Shouldn't rotating a key wipe out the ability to use a new key and require writing new manifest lists and replacing old ones?

I know we can get similar behavior with a Rotate + Expire, but is that a real use case?

Reviewer: I talked a little about this idea with Russell and I think the trade-off is whether to re-encrypt all of the snapshot keys when rotating the table key. If we re-encrypt the snapshot keys, that could take a while when there are potentially thousands of snapshots in the table. On the other hand, in practice it would be a small set and rotating the table key would be rare.

I think I like the idea of a single table key that is used to decrypt all of the snapshot keys. It would require a structure that allows us to change the content of the snapshot keys (a separate list of them that is mutable) but that wouldn't be too difficult.

@ggershinsky, what do you think about that idea?

|            |            | _optional_ | **`key-id`**                | The ID of the encryption key that encrypts the manifest list key metadata. See [Snapshot](#key-metadata-key-id).                                                                                                                                                                                                                                                                                 |

> |            |            | _optional_ | **`key-wrap`**              | Wrapped (encrypted) metadata encryption key. Wrapping can for example be done in a Mey Management Service (KMS).                                                                                                                                                                                                                                                                                 |

Reviewer: Are these intended to be fields inside of `key-cache`? I think that keys in table metadata should be a list of objects and those objects should have a table defining the structure. That's how we've added other structures, like schemas: "A list of schemas, stored as objects with `schema-id`".

Something like this:
```
_optional_ | **`encryption-keys`** | A list of encryption keys, stored as objects. |

...

Each encryption key within the `encryption-keys` table metadata field is a struct with the following fields:
| v1/v2 | v3 | Field name | Type | Description |
|----|----|------------|------|-------------|
|  | _required_ | **`key-id`** | `string` | ID used to refer to this key in table metadata. |
...
```

One thing we can do to make this more flexible is to allow each key to have a key ID that encrypted it, or to indicate that it was wrapped by KMS. That would allow using the same `encryption-keys` list for multiple strategies:
* Wrap each snapshot key using KMS
* Wrap a table key with KMS that is used to encrypt snapshot keys
    * Re-encrypt with the new KMS key when rotating
    * Leave the existing KMS-wrapped key when rotating

and then the table has evolved into `spec#1` which has just one field `{field#2}`.
The unified partition type looks like `Struct<field#1, field#2>`.

#### Encryption Keys

Encryption keys and metadata required for decrypting the manifest list files in encrypted tables.
There are two types of entries:

> 1. `key-metadata`: serialized key-metadata of the encrypted manifest list files. The key-metadata objects include encryption keys and other fields required to decrypt a file. Since these objects are sensitive, the serialized key-metadata byte arrays are encrypted by another key. The encryption is done via the integrity-preserving AES GCM cipher, using the snapshot ID as the AAD (additional authentication data) parameter. The result of the encryption is converted to a string via base64 encoding.

Reviewer 4: Could use "key metadata" (without the dash) outside of the type but not sure here, something like:
```suggestion
1. `key-metadata`: serialized key metadata of an encrypted manifest list file. Key metadata includes encryption keys and other fields required to decrypt a file. As they are sensitive, serialized key metadata byte arrays are encrypted by another key. This encryption is achieved via the integrity-preserving AES GCM cipher, using the snapshot ID as the AAD (additional authentication data) parameter. The result of the encryption is converted to a string via base64 encoding.
```

Reviewer: Also part of the standard encryption spec.

> 2. `key`: the AES GCM key that encrypts the manifest list key-metadata. Since these keys are sensitive, they are wrapped/encrypted in a Key Management Service (KMS), using the table master key. The result of the wrapping is converted to a string via base64 encoding.

Reviewer 4: ```suggestion
2. `key`: the AES GCM key that encrypts the manifest list key-metadata. Since these keys are sensitive, they are encrypted, or "wrapped", by a Key Management Service (KMS), using the table master key. KMS-wrapped keys are converted to strings via base64 encoding.
```

maybe defines "KMS-wrapping" a bit more clearly?

Reviewer: We've removed this from the table spec because it concerns how keys are managed in standard encryption, which will be its own spec.

`encryption-keys` field of table metadata is an optional list of structs with the following fields:

| v1 | v2 |     v3     |     Field name          |   Type.  |                             Description                                                                  |
|----|----|------------|-------------------------|----------|----------------------------------------------------------------------------------------------------------|

> |    |    | _required_ | **`id`**                | `string` | ID of the encryption key.                                                                                |

Author: maybe something like `key-id` for less ambiguity

Reviewer: Used `key-id`.

> |    |    | _required_ | **`type`**              | `string` | `key-metadata` for manifest list encryption key metadata, or `key` for keys that encrypt `key-metadata`. |

Author: the implementation shows that we don't need this field, technically. The keys are simply referenced by a `key-id`. 
Wrt human readability, we can infer the type using the `encryption-key-id` field (it will be filled for a `key-metadata` and empty for a `key`)

> |    |    | _required_ | **`key-metadata`**.     | `string` | Encrypted `key-metadata` or wrapped `key`. Base64-encoded.                                               |

Reviewer 4: Maybe `encrypted-payload` or just `payload`? I think this will always be encrypted (as opposed to the `key_metadata` fields inside manifest lists and manifests) so perhaps nice to reflect that in the name

> |    |    | _required_ | **`key-metadata`**.     | `string` | Encrypted `key-metadata` or wrapped `key`. Base64-encoded.                                               |

Reviewer: Updated to `encrypted-key-metadata`.
